src/systems/filecoin_blockchain/storage_power_consensus/_index.md

```diff
@@ -44,54 +46,124 @@ More precisely,
+    for rand == nil {
+        seeds := GetRandomSeedsForEpoch(epoch)
+        if seeds == nil {
+            epoch -= 1
+        } else {
+            rand = seeds[len(seeds)-1].Randomness
+        }
+    }
+    return rand
+}
+
+GetRandomSeedsForEpoch(epoch) []BeaconEntry {
+    maxDrandRound := MaxBeaconRoundForEpoch(epoch)
+
+    if epoch == 0 {
+        // special case genesis
+        entries := []
+        rand := drand.Public(curr)
+        return append(entries, rand)
+    }
+
+    prevMaxDrandRound := MaxBeaconRoundForEpoch(epoch - 1)
+    if (maxDrandRound == prevMaxDrandRound) {
+        // no new drand randomness
+        return nil
+    }
+
+    entries := []
+    curr := maxDrandRound
+    // fetch all drand entries generated between this epoch and last
+    for curr > prevMaxDrandRound {
+        rand := drand.Public(curr)
+        entries = append(entries, rand)
+        curr -= 1
+    }
+    return entries
+}
+```
 
-- The {{<sref sector_sealer>}} uses tickets as SealSeeds to bind sector commitments to a given subchain.
-- The {{<sref post_generator>}} likewise uses tickets as PoStChallenges to prove sectors remain committed as of a given block.
-- They are drawn by the Storage Power subsystem as randomness in {{<sref leader_election>}} to determine their eligibility to mine a block
-- They are drawn by the Storage Power subsystem in order to generate new tickets for future use.
+{{<label beacon_entry_validation>}}
+### Validating Beacon Entries on block reception
+
+Per the above, a Filecoin chain will contain the entirety of the beacon's output from the Filecoin genesis to the current block.
+
+Given their role in leader election and other critical protocols in Filecoin, a block's beacon entries must be validated for every block. See {{<sref drand>}} for details. This can be done by ensuring every beacon entry is a valid signature over the prior one in the chain, using drand's [`Verify`](https://github.com/drand/drand/blob/master/beacon/beacon.go#L72) endpoint as follows:
+
+```go
+// This need not be done for the genesis block
+// We assume that blockHeader and priorBlockHeader are two valid subsequent headers where block was mined atop priorBlock
+ValidateBeaconEntries(blockHeader, priorBlockHeader) error {
+    currEntries := blockHeader.BeaconEntries
+    prevEntries := priorBlockHeader.BeaconEntries
+
+    var entries []BeaconEntry
+    // at currIdx == 0, must fetch last Fil block's last BeaconEntry
+    entries := append(entries, prevEntries[len(prevEntries)-1])
+    entries := append(entries, currEntries...)
+
+    currIdx := len(entries) - 1
+    for currIdx >= 0 {
+        // walking back the entries to ensure they follow one another
+        currEntry := entries[currIdx]
+        prevEntry := entries[currIdx - 1]
+        err := drand.Verify(node.drandPubKey, prevEntry.Data, currEntry.Data, currEntry.Round)
+        if err != nil {
+            return err
+        }
+        currIdx -= 1
+    }
+
+    return nil
+}
```

Comment:
this seems to verify the entries but not to check the validity with respect to filecoin (e.g. checking that we don't have more entries or less entries for this particular epoch)

Reply:
That's a fair point... I'm not sure how you'd do that without overly constraining (nor why it's strictly necessary)...

@Kubuxu can you point to where this is done in Lotus?